### ISSUE TITLE
RFC: API Design

### ISSUE THREAD

**Issue by tyrannosaurus-becks** [approzium/approzium#75]
## Request For Comment

This Github issue is a Request For Comment regarding adding an administrative API to Approzium.

## Background

Currently, to plant credentials for the Go authenticator to use in creating connections, people must directly place credentials in their storage back-end. They are placed as a JSON map, for example, these are placed for those using HashiCorp Vault:

```
$ vault write approzium/1.2.3.4:5432 dbuser1=@dbuser1-creds.json
# creds.json is: 
{
    "password": "asdfghjkl",
    "iam_arns": [
        "arn:aws:iam::accountid:role/rolename1",
        "arn:aws:iam::accountid:role/rolename2"
    ]
}
```

This approach is brittle. A slight typo on the user's end regarding the Vault path or the credentials will result in the Authenticator failing to locate or authorize use of the credentials.

We would like to add an API that improves user experience.

### At Startup

- We would fail if a credential manager wasn't configured for persistent storage.
- If we didn't detect any administrators, we would provide a message like, "Welcome! Your setup API key is 12345".
  - We would encourage administrators to hit an endpoint like `"/v1/startupkey/revoke"` when finished setting up.
- Using this key, the admin could do first-time setup of administrators, including setting up themself as an admin.

### After Startup Key Revocation

Administrators would use the same authentication methods as machines. For example, supposing my own IAM ARN had been added as an administrator, I would be able to hit all these endpoints. We would create a CLI tool that does the work of pulling my identity from my local environment, and allows me to execute all calls. This would be a simple, downloadable, command-line binary built in Go. It would hit the endpoints outlined below.

### Admin Management Endpoints

We would expose the following endpoints:

- POST (create, update, delete) `/v1/administrators/:name`
  - method: AWS
  - iam_arns: [some-arns]
- GET (list response) `/v1/administrators`
  - // all the people

### Credential Management Endpoints

We would expose endpoints specific to each database type for which we were given credentials. For example, since we currently support postgres:

- POST (create, update, delete) `/v1/postgres/:dbhost/:dbport`
  - username: bob
  - password: foo
  - iam_arns: [foo]
- GET (list response) `/v1/postgres`, `/v1/postgres/:dbhost`

### Endpoint Visibility

We would ensure that all endpoints were well described from the outset using Swagger or OpenAPI (tbd), so that it would be easy in the future for others to build integrations with us, or for a UI to be built.

## Comments/Thoughts Welcome!

This is an early-stage RFC and I am happy to edit it as your feedback arrives! Please don't hesitate to comment. :-)

**Comment by timothydnguyen** [comment]
This is great! One question I have is why does the db type (e.g. `postgres`) need to be part of the cred mgmt endpoint? Is this in case we end up storing in different structures for various db types?

**Comment by tyrannosaurus-becks** [comment]
@timothydnguyen yeah, that's what I was thinking. So maybe for MySQL it would be `/v1/mysql/:dbhost/:dbport`. However, that's a very good question. Do you think it would be better if it were just `/v1/:dbhost/:dbport`? 

On the plus side, it would be shorter. Also, the `:dbhost` field might already be rather descriptive, like it could be, `my-company-mysql`. 

On the minus side, if the `:dbhost` were simply `127.0.0.1` for a bunch of locally accessible databases, it could be hard to reason about which one a cred is for by only looking at the ports.

Another option might be to make the endpoint something more like `/v1/:human-friendly-name`, and its parameters would be dbhost, port, username, password to return, and authorized iam arns.

**Comment by tyrannosaurus-becks** [comment]
@UpGado @timothydnguyen I updated it above per our conversation, let me know if you have further thoughts!
